src/inc/til/u8u16convert.h

```diff
@@ -306,25 +89,117 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
         }
     }
 
+#pragma warning(push)
+#pragma warning(disable : 26429 26446 26459 26481 26482 26485) // use not_null, subscript operator, use span, pointer arithmetic, dynamic array indexing, array to pointer decay
     // Routine Description:
     // - Takes a UTF-8 string, complements and/or caches partials, and performs the conversion to UTF-16.
     // Arguments:
     // - in - UTF-8 string to be converted
     // - out - reference to the resulting UTF-16 string
-    // - state - reference to a til::u8state class holding the status of the current partials handling
+    // - state - reference to a til::u8state holding the status of the current partials handling
     // Return Value:
     // - S_OK          - the conversion succeeded
     // - E_OUTOFMEMORY - the function failed to allocate memory for the resulting string
     // - E_ABORT       - the resulting string length would exceed the upper boundary of an int and thus, the conversion was aborted before the conversion has been completed
     // - E_UNEXPECTED  - an unexpected error occurred
-    template<class inT, class outT>
-    [[nodiscard]] typename std::enable_if<std::is_same<typename inT::value_type, char>::value && std::is_same<typename outT::value_type, wchar_t>::value, HRESULT>::type
-    u8u16(const inT in, outT& out, u8state& state) noexcept
+    template<class outT>
+    [[nodiscard]] HRESULT u8u16(const std::string_view& in, outT& out, u8state& state) noexcept
     {
-        std::string_view sv{};
-        RETURN_IF_FAILED(state(std::string_view{ in }, sv));
-        return til::u8u16(sv, out);
+        try
+        {
+            out.clear();
+            RETURN_HR_IF(S_OK, in.empty());
+
+            int capa16{};
+            // The worst ratio of UTF-8 code units to UTF-16 code units is 1 to 1 if UTF-8 consists of ASCII only.
+            RETURN_HR_IF(E_ABORT, !base::CheckAdd(in.length(), state.have).AssignIfValid(&capa16));
+
+            out.resize(gsl::narrow_cast<size_t>(capa16));
+            auto len8{ gsl::narrow_cast<int>(in.length()) };
+            int len16{};
+            auto cursor8{ in.data() };
+            if (state.have)
+            {
+                if (state.want > len8) // we still didn't get enough data to complete the code point, however this is not an error
+                {
+                    std::move(cursor8, cursor8 + len8, state.partials + state.have);
+                    state.have += gsl::narrow_cast<uint8_t>(len8);
+                    state.want -= gsl::narrow_cast<uint8_t>(len8);
+                    out.clear();
+                    return S_OK;
+                }
+
+                std::move(cursor8, cursor8 + state.want, state.partials + state.have);

```

Comment:
I suspect it's rather unlikely that `if (state.want > len8)` is true.
So how about we write it as if it was wrong most of the time:
```cpp
const auto copyable = std::min<int>(len8, state.want);
memcpy(&state.partials[state.have], cursor8, copyable);
state.have += copyable;
state.want -= copyable;

if (state.want) {
	return S_OK;
}

len16 = MultiByteToWideChar(CP_UTF8, 0, state.partials, state.have, out.data(), capa16);
RETURN_HR_IF(E_UNEXPECTED, !len16);

capa16 -= len16;
len8 -= state.have;
cursor8 += state.have;

// state.want is already zero at this point
state.have = 0;
```

Reply:
While I agree with you that it's rather unlikely, we even have a unit test for code unit by code unit input. It's the result from #4673 where I would have expected to never get only one code unit as input. However, @j4james was facing this bug.

Reply:
I didn't mean to say to remove the if condition, but rather that we can deduplicate the logic that appends characters.